src/feed.sol

```diff
@@ -0,0 +1,159 @@
+    mapping (bytes32 => uint) public maturityDate;
+
+    // risk => recoveryRatePD
+    mapping (uint => uint) public recoveryRatePD;
+
+    // loan => futureValue
+    mapping (uint => uint) public futureValue;
+
+    WriteOff [2] public writeOffs;
+
+    struct WriteOff {
+        uint rateGroup;
+        // denominated in RAY (10^27)
+        uint percentage;
+    }
+
+    uint public discountRate;
+    uint public maxDays;
+
+    constructor (uint discountRate_, uint maxDays_) public {
+        discountRate = discountRate_;
+        maxDays = maxDays_;
+    }
+
+    function init() public {
+        super.init();
+        // gas optimized initialization of writeOffs
+        // write off are hardcoded in the contract instead of init function params
+
+        // risk group recoveryRatePD
+        recoveryRatePD[0] = ONE;
+        recoveryRatePD[1] = 90 * 10**25;
+
+        // 60% -> 40% write off
+        // 91 is a random sample for a rateGroup in pile for overdue loans
+        writeOffs[0] = WriteOff(91, 6 * 10**26);
+        // 80% -> 20% write off
+        // 90 is a random sample for a rateGroup in pile for overdue loans
+        writeOffs[1] = WriteOff(90, 8 * 10**26);
+    }
+
+    function uniqueDayTimestamp(uint timestamp) public pure returns (uint) {
+        return (1 days) * (timestamp/(1 days));
+    }
+
+    /// maturityDate is a unix timestamp
+    function file(bytes32 what, bytes32 nftID_, uint maturityDate_) public {
+        if (what == "maturityDate") {
+            maturityDate[nftID_] = uniqueDayTimestamp(maturityDate_);
+        } else { revert("unknown config parameter");}
+    }
+
+    /// Ceiling Implementation
+    function borrow(uint loan, uint amount) external auth {
+
+        // ceiling check uses existing loan debt
+        require(ceiling(loan) >= safeAdd(pile.debt(loan), amount), "borrow-amount-too-high");
+
+        bytes32 nftID_ = nftID(loan);
+        uint maturityDate_ = maturityDate[nftID_];
+
+        // calculate future value FV
+        uint fv = calcFutureValue(loan, amount, maturityDate_);
+        futureValue[loan] = safeAdd(futureValue[loan], fv);
+
+        if (buckets[maturityDate_].value == 0) {
+            addBucket(maturityDate_, fv);
+            return;
+        }
+
+        buckets[maturityDate_].value = safeAdd(buckets[maturityDate_].value, fv);
+    }
+
+    function calcFutureValue(uint loan, uint amount, uint maturityDate_) public returns(uint) {
+        return rmul(rmul(rpow(pile.loanRates(loan),  safeSub(maturityDate_, uniqueDayTimestamp(now)), ONE), amount), recoveryRatePD[risk[nftID(loan)]]);
+
+    }
+
+    function repay(uint loan, uint amount) external auth {
+        uint maturityDate_ = maturityDate[nftID(loan)];
+
+        // remove future value for loan from bucket
+        buckets[maturityDate_].value = safeSub(buckets[maturityDate_].value, futureValue[loan]);
+
+        uint debt = pile.debt(loan);
+
+        debt = safeSub(debt, amount);
```

Comment:
amount can be bigger then debt. I would check if amount is higher then debt -> debt = 0

Reply:
safeSub would always revert amount is higher then debt 🙈

Reply:
It is guaranteed the amount is always <= debt. I wanted to avoid the double check.

It happens here: https://github.com/centrifuge/tinlake/blob/develop/src/borrower/shelf.sol#L206

Reply:
true